### PR TITLE
Always ensure /dev/{u,random} are in initramfs

### PR DESCRIPTION
**Pull request by cgwalters**
In some scenarios, such as running `dracut` inside a podman/docker container, we may not have the capabilities necessary to create devices.

But we don't need the ability to make devices; instead, just carry a static copy of a small cpio archive with that content pre-created, and ensure it's the first thing in the generated initramfs image.

EDIT: Downstream in RHEL this is https://bugzilla.redhat.com/show_bug.cgi?id=2208040

### CODE REVIEW

**Comment by cgwalters** [comment]
(draft since this could use a bit more testing)

**Comment by cgwalters** [comment]
The originating bug here is https://issues.redhat.com/browse/OCPBUGS-13207
But we shipped prior workarounds like https://github.com/coreos/rpm-ostree/pull/4058 in our wrapper for dracut.

The rationale for this is basically "server side" initramfs image builds in derived images.

**Comment by cgwalters** [comment]
Yes.  For more information, see https://fedoraproject.org/wiki/Changes/OstreeNativeContainer and https://coreos.github.io/rpm-ostree/container/ - we use containers as a build tool to generate bootable host systems.  (As of recently there's also https://github.com/containers/bootc which streamlines this)

**Comment by cgwalters** [comment]
> Perhaps to move this forward try to keep dracut-random.cpio.gz out of upstream and make only the necessary changes that would allow you to carry dracut-random.cpio.gz for your own needs. Would you see this as a good compromise ?

It'd help for sure, but I wouldn't call it a good compromise.  What *is* the compromise here - i.e. what do you see as bad?  Just having additional code?  The opacity of the pre-generated file (that's definitely fixable)?  Something else?

**Comment by cgwalters** [comment]
This keeps coming up (e.g. https://github.com/coreos/rpm-ostree/issues/4499 ), and if we don't agree to merge this, what I'll probably end up doing is adding a wrapper around `dracut` that does it anyways.  We actually already have infrastructure for this.  It's just much less sustainable than fixing the initramfs to just have the devices by default.

**Comment by cgwalters** [comment]
We could try making the cpio file at build time, but that's going to have a potentially larger blast radius - it's quite likely some build systems may actually be using container runtimes with similar device cgroup restrictions.

Another possibility is to try to land a patch in upstream `cpio` to allow it to synthesize these devices directly.

Yet another possibility is to generate a cpio blob in `dracut-install` in C code.

**Comment by pvalena** [comment]
Hello,

yes, I did a brief investigation into this PR & the cpio archive is the main obstackle in bringing this even to RHEL-9 as well. 

Unfortunately, I didn't have spare cycles yet to look into what's the archive content and whether it can be recreated at dracut runtime (apart from copied from the system).

**Comment by ddiss** [comment]
> Is this "cpio blob" portable across architectures (x86, x86_64, arm, arm64, etc) or they they need to be generated separate for each architecture ?

If the archive blob is just the ` /dev/{u,}random` nodes, then it shouldn't be arch specific.

I agree that it'd be nice to avoid carrying the blob in dracut. There are a couple of other options which could be considered here:
If the use-case assumes full control over the kernel, then the payload could easily be nested in the kernel via `CONFIG_INITRAMFS_SOURCE`, or a minor change to `default_cpio_list`:
```
--- a/usr/default_cpio_list
+++ b/usr/default_cpio_list
@@ -4,3 +4,5 @@
 dir /dev 0755 0 0
 nod /dev/console 0600 0 0 c 5 1
 dir /root 0700 0 0
+nod /dev/random 0666 0 0 c 1 8
+nod /dev/urandom 0666 0 0 c 1 9
```

Still, the underlying problem here is the source-directory-tree-defines-archive-contents interface for GNU cpio (which is copied for dracut-cpio, which breaks unprivileged archive creation. My plan is to address this with https://github.com/dracutdevs/dracut/issues/1662 , which will add a Linux `gen_init_cpio` like interface to dracut-cpio.

**Review by ddiss — CHANGES_REQUESTED** [review]

File: Makefile [hunk 1]
```diff
@@ -207,6 +207,7 @@ endif
 	if [ -f dracut-util ]; then \
 		install -m 0755 dracut-util $(DESTDIR)$(pkglibdir)/dracut-util; \
 	fi
+	install -m 0644 -Dt $(DESTDIR)$(pkglibdir) dracut-random.cpio.gz
```
[the remark this answers is not in the document]
Comment (ddiss) [reply Makefile:210]: One thing I forgot to mention: GNU cpio unnecessarily zero-pads archives out to the next "block-size" (default=512) alignment, while the newc / SVR4 spec only requires 4-byte alignment, so any bundled archive with GNU cpio padding could be trimmed down.
In matching GNU cpio quirks, dracut-cpio does the same:
https://github.com/dracutdevs/dracut/blob/master/src/dracut-cpio/src/main.rs#L644

**Comment by cgwalters** [comment]
Closing since we're in agreement this exact patch isn't the way to go

**Comment by ericcurtin** [comment]
> Is this "cpio blob" portable across architectures (x86, x86_64, arm, arm64, etc) or they they need to be generated separate for each architecture ?

Note, it's important that the solution is also portable across compressors/decompressors (gzip, lz4, zstd, etc.) Encountered an issue where part of the initramfs was lz4 and another part gzip and the kernel decompressor got confused and panicked on boot.

**Pull request closed without merging**